Add intrinsic support under onAcc namespace

LukeTheWalker opened the pull request:
## Summary
This PR introduces a portable `popcount` (population count) intrinsic available via the `alpaka::onAcc` namespace. This allows users to count the number of set bits in an integer type across all supported backends (CPU, CUDA, HIP, and SYCL).

## Changes

- **New API:** Added `alpaka::onAcc::popcount(acc, value)` which dispatches to the correct backend implementation.
- **Trait Infrastructure:** Introduced `GetIntrinsicImpl` and specific tag types (`StlIntrinsic`, `CudaHipIntrinsic`, `SyclIntrinsic`) to handle intrinsic dispatching based on the active executor.
- **Backend Implementations:**
  - **Host (Serial, OMP, TBB):** Utilizes C++20 `std::popcount` from `<bit>`. Includes handling for signed/unsigned casting.
  - **CUDA / HIP:** Maps to `__popc` (for 32-bit) and `__popcll` (for 64-bit).
  - **SYCL / oneAPI:** Maps to `sycl::popcount`.
- **Testing:** Added a new unit test `tests/unit/intrinsic/popc.cpp` verifying correctness against a reference implementation on all enabled backends.

LukeTheWalker committed:
feat: added popcount support under onAcc namespace

refactor: moved the popc example into a unit test

LukeTheWalker commented:
Closes #368

psychocoderHPC commented on include/alpaka/api/cuda/executor.hpp:
```diff
@@ -44,6 +44,14 @@ namespace alpaka::onAcc::trait
+    template<>
+    struct GetIntrinsicImpl::Op<alpaka::exec::GpuCuda>
+    {
+        constexpr decltype(auto) operator()(alpaka::exec::GpuCuda const) const
```
could you change `constexpr decltype(auto)` to `constexpr int` that the user knows what is the result. It would than follow `std::popcount()`

LukeTheWalker committed:
fix: readded template<> in template specialization

psychocoderHPC commented:
@LukeTheWalker Thanks for you pullrequest, there are only few small things which should be changed, I will push the change as PR to your fork. 
You opened the PR from your dev branch, therefore I do not want to push directly on top of your dev branch.
After you accept the PR it will be visible in this PR.

I will ping you as soon I am finished.

psychocoderHPC committed:
refactor `popCount` PR

- switch internal intrincic namespace to `onAcc::internal::intrinic`
- move STL intrincics to `onAcc/internal`
- do not use `#include <alpaka/...>` within alpaka itself

psychocoderHPC commented:
@LukeTheWalker I opened the PR https://github.com/LukeTheWalker/alpaka3-intrinsics/pull/1

I moved some code fromthe  generic alpaka folder into `onAcc` because we try to collect all in kernel function under `onAcc`.
For the test case I used simply `std::pop_count`.
We include within `include/alpaka` everything with `#include "alpaka/..."` to not break single header usage. see the godbolt example https://alpaka3.readthedocs.io/en/latest/basic/example.html#play-online-with-compiler-explorer
I run auto formation: https://alpaka3.readthedocs.io/en/latest/contribution/tools.html#pre-commit what removed some spaces, ...

LukeTheWalker commented:
@psychocoderHPC Thanks.

Sorry for not running the pre-commit, I started the branch as a personal thing and then forgot to check the contributions page.

psychocoderHPC commented:
> @psychocoderHPC Thanks.
> 
> Sorry for not running the pre-commit, I started the branch as a personal thing and then forgot to check the contributions page.

no problem, big thanks for your contribution.

psychocoderHPC committed:
fix compile ...

due to wrong namespace

psychocoderHPC commented:
@LukeTheWalker I puhed this time directly onto you branch to fix a tiny style issue and a wrong namespace I introduced.

psychocoderHPC commented:
@LukeTheWalker, I was thinking about the requirement of the `acc` in the intrinsic function. I came to the conclusion that this is not required equally to math functions. This provides the possibility to use such functions e.g. within class operators where the accelerator is typically not available.

I opened PR #371 which includes this PR. This PR will be merged first and after that #371 to build the base for more intrinsics e.g. `ffs`.
